### ISSUE TITLE
[BUG] Different Batch Size for Train and Validation data. 

### ISSUE THREAD

**Issue by XericZephyr** [dmlc/mxnet#648]
# Bug Description

There is a bug found when the `batch_size` of train and validation data is different. 

The behavior is actually very weird. 

1. When the train `batch_size` < val `batch_size`, the validation accuracy will always be zero. 
2. When the train `batch_size` == val `batch_size`, the result is fine. 
3. When the train `batch_size` > val `batch_size`, there is an exception raised. 

I think the main reason is actually due to the `ndarray.copyto()` implementation in `eval_data` loading.
Context: https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L358 

1. When the source is larger than the destination, it will **silently** truncate the source `ndarray`. 
2. When the source is smaller than the destination, it reports an error. 

I think it's unsafe to silently do this truncation. 

# Solution 

1. We change the behavior of ndarray, at least we should warn the user when the size of the two ends of the copy procedure mismatch. I think a check in https://github.com/dmlc/mxnet/blob/master/python/mxnet/ndarray.py#L346 is enough, probably.  
2. We warn the user when we found the `batch_size` of actual label and prediction label is different, for example in https://github.com/dmlc/mxnet/blob/master/python/mxnet/metric.py#L53 . 
We `assert pred.shape[0] == label.shape[0], "Batch size of train and validation data should match."`

**Comment by XericZephyr** [comment]
I will try to contribute a temporary solution to fix the case when train `batch_size` < val `batch_size` for Accuracy metric. It will come shortly.

**Comment by piiswrong** [comment]
I think It would be better to just force them to be identical, since the network is only allocated for training batchsize not testing

**Comment by XericZephyr** [comment]
I think so. Just forcing them to be the same is more rigid. But for who is not familiar with mxnet framework, this kind of restriction is very frustrating. For now, I think allowing user to set train `batch_size` <= val `batch_size` might leave user more flexibility. 

So, here is the choice, the rigid one or user-friendly one?

**Comment by tqchen** [comment]
merged in the fix by @XericZephyr for now. Thanks! In future, enforcing the same batch size might be a better idea

**Comment by Tiamaty** [comment]
Hi,
I would like to know if you guys really enforced the batch_size and the val_batch_size to be the same size (in ipython notebook the kernel dies, maybe a seg fault ?).
I am asking because from a theoretical point of view if you enforce that the users wont be able to use cross-validation methods like K-Fold or Jack Knife where the amount of data in validation is smaller than in training.

**Comment by Tiamaty** [comment]
Just to clarify, I ran across this while doing a cross-validation using mxnet. When the number of folds is 2 meaning that the amount of data in training and validation is equal everything is fine but if I raise the number of folds the kernel dies.

**Comment by Tiamaty** [comment]
Any comments on these ?
@tqchen @XericZephyr
